### PR TITLE
kubectl bind / kubectl bind apiservice split

### PR DESCRIPTION
**Pull request by sttts**
Including e2e.

### CODE REVIEW

File: contrib/example-backend/controllers/servicebindingrequest/servicebindingresource_reconcile.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+				GroupResource: res.GroupResource,
+			})
+		}
+		logger.V(2).Info("Creating service export", "namespace", export.Namespace, "name", export.Name)
+		var err error
+		if _, err = r.createServiceExport(ctx, export); err != nil && !apierrors.IsAlreadyExists(err) {
+			export.GenerateName = req.Name + "-"
+			export.Name = ""
+
+			logger.V(2).Info("Creation of service export failed, trying with a generated name", "namespace", export.Namespace, "name", export.Name)
```
Comment (s-urbaniak) [inline contrib/example-backend/controllers/servicebindingrequest/servicebindingresource_reconcile.go:78]: what's the benefit of this?

Comment (s-urbaniak) [reply contrib/example-backend/controllers/servicebindingrequest/servicebindingresource_reconcile.go:78]: i.e. vs. always creating with a generate name

File: contrib/example-backend/controllers/servicebindingrequest/servicebindingresource_reconcile.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+		}
+		if err != nil {
+			return err
+		}
+
+		req.Status.Export = export.Name // waiting for export to be ready for phase update
+		return nil
+	}
+
+	export, err := r.getServiceExport(req.Namespace, req.Status.Export)
```
Comment (s-urbaniak) [inline contrib/example-backend/controllers/servicebindingrequest/servicebindingresource_reconcile.go:91]: if this was created using a generate name, `req.Status.Export` will still point to the non-generate one.

Comment (s-urbaniak) [reply contrib/example-backend/controllers/servicebindingrequest/servicebindingresource_reconcile.go:91]: nevermind, linke 87 updates it 👍

File: contrib/example-backend/controllers/servicebindingrequest/servicebindingresource_reconcile.go (hunk 1 above)
Comment (sttts) [reply contrib/example-backend/controllers/servicebindingrequest/servicebindingresource_reconcile.go:78]: The APIServiceBindingRequest has ann optional name in the authn response. This is supposed to be used if the service provider keeps hygiene in cluster namespaces. But this cannot be guaranteed. Hence, this special case.

File: contrib/example-backend/controllers/servicebindingrequest/servicebindingresource_reconcile.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+			logger.V(2).Info("Creation of service export failed, trying with a generated name", "namespace", export.Namespace, "name", export.Name)
+			if _, err = r.createServiceExport(ctx, export); err != nil {
+				return err
+			}
+		}
+		if err != nil {
+			return err
+		}
+
+		req.Status.Export = export.Name // waiting for export to be ready for phase update
```
[the remark this answers is not in the document]
Comment (sttts) [reply contrib/example-backend/controllers/servicebindingrequest/servicebindingresource_reconcile.go:87]: fixed

File: contrib/example-backend/controllers/servicebindingrequest/servicebindingresource_reconcile.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+		return nil
+	}
+
+	export, err := r.getServiceExport(req.Namespace, req.Status.Export)
+	export = export.DeepCopy()
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	} else if apierrors.IsNotFound(err) {
+		conditions.MarkFalse(
+			export,
```
Comment (s-urbaniak) [inline contrib/example-backend/controllers/servicebindingrequest/servicebindingresource_reconcile.go:97]: shouldn't this be `req` ?

Comment (sttts) [reply contrib/example-backend/controllers/servicebindingrequest/servicebindingresource_reconcile.go:97]: it shoulds, and explains the race previously

File: contrib/example-backend/controllers/servicebindingrequest/servicebindingresource_reconcile.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+			kubebindv1alpha1.APIServiceBindingRequestConditionExportReady,
+			"Unknown",
+			conditionsapi.ConditionSeverityInfo,
+			"APIServiceExport %s in the service provider cluster has no Ready condition",
+			req.Status.Export,
+		)
+		return nil
+	} else if c.Status != corev1.ConditionTrue {
+		conditions.MarkFalse(
+			export,
```
Comment (s-urbaniak) [inline contrib/example-backend/controllers/servicebindingrequest/servicebindingresource_reconcile.go:119]: shouldn't this be `req`?

File: contrib/example-backend/controllers/servicebindingrequest/servicebindingresource_reconcile.go [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+			conditionsapi.ConditionSeverityError,
+			"APIServiceExport %s in the service provider cluster not found",
+			req.Status.Export,
+		)
+		return nil
+	}
+
+	if c := conditions.Get(export, conditionsapi.ReadyCondition); c == nil {
+		conditions.MarkFalse(
+			export,
```
[the remark this answers is not in the document]
Comment (sttts) [reply contrib/example-backend/controllers/servicebindingrequest/servicebindingresource_reconcile.go:109]: done

File: contrib/example-backend/controllers/servicebindingrequest/servicebindingresource_reconcile.go (hunk 5 above)
Comment (sttts) [reply contrib/example-backend/controllers/servicebindingrequest/servicebindingresource_reconcile.go:119]: done

File: pkg/apis/kubebind/v1alpha1/apiservicebindingrequest_types.go [hunk 7]
```diff
@@ -76,13 +82,16 @@ func (in *APIServiceBindingRequest) SetConditions(conditions conditionsapi.Condi
 type APIServiceBindingRequestSpec struct {
 	// parameters holds service provider specific parameters for this binding
 	// request.
+	//
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="parameters are immutable"
```
Comment (s-urbaniak) [inline pkg/apis/kubebind/v1alpha1/apiservicebindingrequest_types.go:94]: i believe this is optional? If so, we need a top-level CEL a la `// +kubebuilder:validation:XValidation:rule="!has(oldSelf. parameters) || has(self. parameters)", message="parameters are required once set"` to disallow removal, see https://kubernetes.io/blog/2022/09/29/enforce-immutability-using-cel/#immutablility-after-first-modification.

Comment (s-urbaniak) [reply pkg/apis/kubebind/v1alpha1/apiservicebindingrequest_types.go:94]: this would be a CEL on the `APIServiceBindingRequestSpec` struct.

Comment (sttts) [reply pkg/apis/kubebind/v1alpha1/apiservicebindingrequest_types.go:94]: oh fun. Thank for pointing that out.

**Review by s-urbaniak — APPROVED** [review]

**Review by s-urbaniak — APPROVED** [review]

**Pull request merged**